README.md
# OpenRCA UI

[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/3912/badge)](https://bestpractices.coreinfrastructure.org/projects/3912)
[![License](https://img.shields.io/github/license/openrca/orca)](https://github.com/openrca/orca-ui)
[![Gitter](https://img.shields.io/gitter/room/openrca/community)](https://gitter.im/openrca/community)

This repository contains the source code for the OpenRCA web interface.

|              Cluster View              |             Cluster Filtering                  |             Alerts View                |
| :------------------------------------: | :--------------------------------------------: | :------------------------------------: |
| ![Graph View](./docs/graph-view.png)   | ![Graph Filtering](./docs/graph-filtering.png) | ![Alerts View](./docs/alerts-view.png) |

## Usage

### Build and run in Docker

Build Docker image:

```sh
docker build -t openrca/orca-ui .
```

Run Docker container (specify correct IP and port of API container):

```sh
docker run -it \
    --name orca-ui \
    -e "BACKEND_URL=http://172.17.0.2:5000" \
    -p 80:80 \
    openrca/orca-ui
```

### Build and run locally

Install dependencies:
```sh
yarn
```

Build:
```sh
yarn build
```

Run:
```sh
yarn start
```

## Development

Navigate to `utils` directory.
Build and start containers by executing command below (requres `docker-compose`):

```sh
docker-compose up
```

Orca UI is now available at [http://localhost](http://localhost).

The code can be modified locally as it is bound to running container.
It will be recompiled on the fly.
Hot reload feature is also enabled, so any changes should be
visible without reloading the webpage.

## Contact

Reach project contributors via these channels:

-   [Gitter chat room](https://gitter.im/openrca/community)
-   [Github issues](https://github.com/openrca/orca-ui/issues)


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/openrca/orca-ui/tree/HEAD)

## [0.2.0](https://github.com/openrca/orca/tree/0.2.0) (2020-09-14)

[Full Changelog](https://github.com/openrca/orca-ui/compare/65c7fb53784a8ebdecbdfce5053b51eb7b5ecfa2...0.2.0)

**Implemented enhancements:**

- Limit object types in default graph view [\#89](https://github.com/openrca/orca-ui/issues/89)
- Include time point argument in requests to Graph API [\#84](https://github.com/openrca/orca-ui/issues/84)
- Display graph as main page [\#81](https://github.com/openrca/orca-ui/issues/81)
- Decrease graph force [\#75](https://github.com/openrca/orca-ui/issues/75)
- Include cluster and node objects in namespace view [\#74](https://github.com/openrca/orca-ui/issues/74)
- Customize alerts table [\#72](https://github.com/openrca/orca-ui/issues/72)
- Render table from node data [\#70](https://github.com/openrca/orca-ui/issues/70)
- Stretch details box to full height [\#69](https://github.com/openrca/orca-ui/issues/69)
- Add object type selector [\#66](https://github.com/openrca/orca-ui/issues/66)
- Customize alerts table [\#60](https://github.com/openrca/orca-ui/issues/60)
- Switch to dark mode [\#50](https://github.com/openrca/orca-ui/issues/50)
- Add on-demand graph reloading [\#40](https://github.com/openrca/orca-ui/issues/40)
- Extend graph payload in Mock API [\#39](https://github.com/openrca/orca-ui/issues/39)
- Highlight alerts [\#38](https://github.com/openrca/orca-ui/issues/38)
- Add graph stats [\#37](https://github.com/openrca/orca-ui/issues/37)
- Update browser title [\#36](https://github.com/openrca/orca-ui/issues/36)
- Add CORS headers to Mock API [\#34](https://github.com/openrca/orca-ui/issues/34)
- Add graph loading animation [\#31](https://github.com/openrca/orca-ui/issues/31)
- Add graph filtering by app namespace [\#30](https://github.com/openrca/orca-ui/issues/30)
- Add alerts filtering [\#29](https://github.com/openrca/orca-ui/issues/29)
- Add alerts list [\#28](https://github.com/openrca/orca-ui/issues/28)
- Add datetime picker [\#26](https://github.com/openrca/orca-ui/issues/26)
- Add object details box [\#25](https://github.com/openrca/orca-ui/issues/25)
- Add graph centering and rescaling [\#22](https://github.com/openrca/orca-ui/issues/22)
- Create production Dockerfile [\#19](https://github.com/openrca/orca-ui/issues/19)
- Improve graph navigation [\#18](https://github.com/openrca/orca-ui/issues/18)
- Update orca-mock API with CORS headers [\#16](https://github.com/openrca/orca-ui/issues/16)
- Load graph structure from Mock API [\#13](https://github.com/openrca/orca-ui/issues/13)
- Prepare Docker compose for development [\#12](https://github.com/openrca/orca-ui/issues/12)
- Dockerize the UI [\#9](https://github.com/openrca/orca-ui/issues/9)
- Implement UI MVP [\#7](https://github.com/openrca/orca-ui/issues/7)
- Mock graph API [\#6](https://github.com/openrca/orca-ui/issues/6)
- Add styling to graph nodes [\#5](https://github.com/openrca/orca-ui/issues/5)
- Add sample graph visualization using D3.js [\#3](https://github.com/openrca/orca-ui/issues/3)

**Fixed bugs:**

- Attribute transform error for large graph [\#91](https://github.com/openrca/orca-ui/issues/91)
- Invalid root for alerts payload [\#86](https://github.com/openrca/orca-ui/issues/86)
- Double icons [\#64](https://github.com/openrca/orca-ui/issues/64)
- Alerts not visible after namespace filtering [\#53](https://github.com/openrca/orca-ui/issues/53)
- Fix date picker position [\#52](https://github.com/openrca/orca-ui/issues/52)

**Merged pull requests:**

- Bump http-proxy from 1.18.0 to 1.18.1 [\#99](https://github.com/openrca/orca-ui/pull/99) ([dependabot[bot]](https://github.com/apps/dependabot))
- Add UI screens to README [\#98](https://github.com/openrca/orca-ui/pull/98) ([bzurkowski](https://github.com/bzurkowski))
- Attribute transform large graph error [\#97](https://github.com/openrca/orca-ui/pull/97) ([k-jano](https://github.com/k-jano))
- Horizontal pod autoscaler icon [\#96](https://github.com/openrca/orca-ui/pull/96) ([filwie](https://github.com/filwie))
- Labels [\#95](https://github.com/openrca/orca-ui/pull/95) ([filwie](https://github.com/filwie))
- Cluster, node, pod and alert as default type [\#94](https://github.com/openrca/orca-ui/pull/94) ([k-jano](https://github.com/k-jano))
- Render a table in detail card, switch type and name atributes [\#93](https://github.com/openrca/orca-ui/pull/93) ([k-jano](https://github.com/k-jano))
- Adjust graph force [\#88](https://github.com/openrca/orca-ui/pull/88) ([bzurkowski](https://github.com/bzurkowski))
- Fix alerts root [\#87](https://github.com/openrca/orca-ui/pull/87) ([bzurkowski](https://github.com/bzurkowski))
- Use timestamp param in graph endpoint [\#85](https://github.com/openrca/orca-ui/pull/85) ([bzurkowski](https://github.com/bzurkowski))
- Customize alerts table [\#83](https://github.com/openrca/orca-ui/pull/83) ([k-jano](https://github.com/k-jano))
- Graph as a main route [\#82](https://github.com/openrca/orca-ui/pull/82) ([k-jano](https://github.com/k-jano))
- Change styling of bottom navbar to fit with the rest of the app \(\#68\) [\#80](https://github.com/openrca/orca-ui/pull/80) ([filwie](https://github.com/filwie))
- Linter fixes in alerts dashboard [\#79](https://github.com/openrca/orca-ui/pull/79) ([k-jano](https://github.com/k-jano))
- Object type multi selector [\#78](https://github.com/openrca/orca-ui/pull/78) ([k-jano](https://github.com/k-jano))
- Display cluster and proper nodes objects in namespace view  [\#77](https://github.com/openrca/orca-ui/pull/77) ([k-jano](https://github.com/k-jano))
- Add node icons [\#63](https://github.com/openrca/orca-ui/pull/63) ([filwie](https://github.com/filwie))
- Alert dashboard [\#62](https://github.com/openrca/orca-ui/pull/62) ([k-jano](https://github.com/k-jano))

src/components/Graph.js
import React from 'react';
import * as d3 from 'd3';
import axios from 'axios';
import Loader from 'react-loader-spinner';
import { Button } from 'react-bootstrap';

import { DateTimePicker } from './DateTimePicker';
import { NodeDetailCard } from './NodeDetailCard';
import { IconMap } from './IconMap';
import './Graph.scss';

function truncate(fullStr, strLen, separator='...') {
  if (fullStr.length <= strLen) return fullStr;

  const sepLen = separator.length,
    charsToShow = strLen - sepLen,
    frontChars = Math.ceil(charsToShow / 2),
    backChars = Math.floor(charsToShow / 2);

  return fullStr.substr(0, frontChars) +
    separator +
    fullStr.substr(fullStr.length - backChars);
}

export class Graph extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      link: null,
      nodeCircle: null,
      nodeIcon: null,
      nodeGroup: null,
      simulation: null,
      namespaceOptions: [],
      objectKindOptions: [],
      showLabels: false,
      namespace: null,
      kinds: ['cluster', 'node', 'pod', 'alert'],
      data: null,
      svg: null,
      g: null,
      loading: true,
      stat: null
    };

    this.zoom = d3.zoom();

    this.generateGraph = this.generateGraph.bind(this);
    this.onDateTimeSelect = this.onDateTimeSelect.bind(this);
    this.ticked = this.ticked.bind(this);
    this.handleNamespaceChange = this.handleNamespaceChange.bind(this);
    this.handleKindChange = this.handleKindChange.bind(this);
    this.nodeCircleRadius = 16;
    this.nodeIconFontSize = 16;
    this.nodeLabelFontSize = 16;
    this.nodeLabelLength = 18;
    this.nodeDetailCard = React.createRef();
    this.scaleGraph = this.scaleGraph.bind(this);
    this.graphLoad = this.graphLoad.bind(this);
    this.handleStatButton = this.handleStatButton.bind(this);
    this.toggleNodeLabels = this.toggleNodeLabels.bind(this);
  }

  componentDidMount() {
    const div = document.getElementById('chart-area');
    div.style.visibility = 'hidden';
    this.prepareSvg();
    this.loadData();
  }

  onDateTimeSelect(timestamp) {
    this.loadData(timestamp);
  }

  toggleNodeLabels(e) {
    this.setState({showLabels: !this.state.showLabels});
    d3.selectAll('.node-label').classed('hidden', this.state.showLabels);
  }

  handleNamespaceChange(e) {
    const namespace = e ? e.value : null;
    this.setState({
      namespace: namespace
    }, () => {
      this.generateGraph(this.state.data);
    });
  }

  handleKindChange(selectList) {
    console.log(selectList);
    const kindList = selectList && selectList.length > 0 ? selectList.map(e => e.value) : null;
    this.setState({
      kinds: kindList
    }, () => {
      this.generateGraph(this.state.data);
    });
  }

  loadData(timestamp = null) {
    const param = timestamp ? `?time_point=${timestamp}` : '';
    axios.get(process.env.REACT_APP_BACKEND_HOST + '/v1/graph' + param)
      .then((response) => {
        const namespaces = [...new Set(response.data.nodes.map(nodeGroup => nodeGroup.properties ? nodeGroup.properties.namespace : null))];
        const namespaceOptions = namespaces.map(namespace => ({
          value: namespace,
          label: namespace
        }));

        const objectTypes = [...new Set(response.data.nodes.map(nodeGroup => nodeGroup.kind))];
        const objectKindOptions = objectTypes.map(type => ({
          value: type,
          label: type
        }));

        this.setState({
          namespaceOptions: namespaceOptions,
          objectKindOptions: objectKindOptions,
          data: response.data
        }, () => {
          if(!timestamp) setTimeout(this.graphLoad, 2000);
          this.generateGraph(response.data);
        });
      })
      .catch((err) => {
        console.log(err);
      });
  }

  clearClicked() {
    d3.selectAll('.clicked').classed('clicked', false);
  }

  nodeMouseOver(nodeGroup) {
    const multiplier = 1.5;
    nodeGroup.selectAll('circle').attr('r', this.nodeCircleRadius * multiplier);
    nodeGroup.selectAll('.node-icon').attr('font-size', `${this.nodeIconFontSize * multiplier}px`);
    nodeGroup.selectAll('.node-label').attr('font-size', `${this.nodeLabelFontSize * multiplier}px`)
      .attr('y', this.nodeCircleRadius * 2 * multiplier)
      .text((d) => d.properties.name);
  }

  nodeMouseOut(nodeGroup) {
    nodeGroup.selectAll('circle').attr('r', this.nodeCircleRadius);
    nodeGroup.selectAll('.node-icon').attr('font-size', `${this.nodeIconFontSize}px`);
    nodeGroup.selectAll('.node-label').attr('font-size', `${this.nodeLabelFontSize}px`)
      .attr('y', this.nodeCircleRadius * 2)
      .text((d) => truncate(d.properties.name, this.nodeLabelLength));
  }

  nodeClick(nodeGroup, nodeData) {
    this.clearClicked();
    nodeGroup.classed('clicked', true);
    this.nodeDetailCard.current.updateNodeData(nodeData);
    this.nodeDetailCard.current.show();
  }

  prepareSvg() {
    const svg = d3.select('#chart-area')
      .append('svg')
      .attr('class', 'graph')
      .style('width', '100%')
      .style('height', '100%');

    const g = svg
      .call(this.zoom.on('zoom', () => {
        g.attr('transform', d3.event.transform);
      }))
      .append('g');

    const simulation = d3.forceSimulation()
      .force('link', d3.forceLink().id(d => d.id).distance(100).strength(1))
      .force('charge', d3.forceManyBody().strength(-300))
      .force('collide', d3.forceCollide().strength(.7).radius(function(d) {
        return d.radius;
      }))
      .force('x', d3.forceX())
      .force('y', d3.forceY())
      .on('tick', this.ticked);

    const link = g.append('g')
      .selectAll('line');

    const nodeGroup = g.append('g')
      .selectAll('.node-group');

    const nodeCircle = g.append('g')
      .selectAll('circle');

    const nodeIcon = g.append('g')
      .selectAll('text');

    const nodeLabel = g.append('g')
      .selectAll('text');

    this.setState({
      link: link,
      nodeGroup: nodeGroup,
      nodeCircle: nodeCircle,
      nodeIcon: nodeIcon,
      nodeLabel: nodeLabel,
      showLabels: false,
      simulation: simulation,
      svg: svg,
      g: g
    });
  }

  graphLoad() {
    this.setState({
      loading: false
    }, () => {
      const div = document.getElementById('chart-area');
      div.style.visibility = 'visible';
      this.scaleGraph();
    });
  }

  scaleGraph() {
    const svg = this.state.svg;
    const g = this.state.g;

    const width = svg.node().getBoundingClientRect().width;
    const height = svg.node().getBoundingClientRect().height;

    const graphX = g.node().getBBox().x;
    const graphY = g.node().getBBox().y;
    const graphWidth = g.node().getBBox().width;
    const graphHeight = g.node().getBBox().height;

    if(graphWidth === 0 || graphHeight === 0) return;

    const scale = 0.95 / Math.max(graphWidth / width, graphHeight / height);

    const transform = d3.zoomIdentity
      .translate(width / 2 - scale * (graphX + graphWidth / 2), height / 2 - scale * (graphY + graphHeight / 2))
      .scale(scale);

    svg
      .transition()
      .duration(0)
      .call(this.zoom.transform, transform);
  }

  filterAlertOtherNamespace(nodes, links) {
    return nodes.filter(nodeGroup => nodeGroup.kind === 'alert').filter(alert => {
      var valid = false;
      links.forEach(link => {
        if(link.source === alert.id || link.target === alert.id) valid = true;
      });
      return !valid;
    });
  }

  filterK8sNodesOtherNamespace(nodes, links) {
    return nodes.filter(nodeGroup => nodeGroup.kind === 'node').filter(node => {
      var valid = false;
      const clusterNames = nodes.filter(nodeGroup => nodeGroup.kind === 'cluster').map(cluster => cluster.id);
      const linksWithoutClusters = links.filter(link => !(clusterNames.includes(link.source) || clusterNames.includes(link.target)));

      linksWithoutClusters.forEach(link => {
        if(link.source === node.id || link.target === node.id) valid = true;
      });
      return !valid;
    });
  }

  generateGraph(data) {
    var links = data.links;
    var nodes = d3.values(data.nodes);

    //Filter by namespace
    if (this.state.namespace) {
      nodes = nodes.filter(nodeGroup => nodeGroup.properties.namespace === this.state.namespace || nodeGroup.kind === 'alert' ||
        nodeGroup.kind === 'cluster' || nodeGroup.kind === 'node');
    }

    var nodesName = nodes.map(nodeGroup => nodeGroup.id);
    links = links.filter(link => nodesName.includes(link.source) && nodesName.includes(link.target));

    //Filter alerts
    const alertsOtherNamespace = this.filterAlertOtherNamespace(nodes, links);
    nodes = nodes.filter(nodeGroup => !alertsOtherNamespace.includes(nodeGroup));

    //Filter nodes
    if(this.state.namespace !== '{}'){
      const k8sNodesOtherNamespace = this.filterK8sNodesOtherNamespace(nodes, links);
      const k8sNodesOtherNamespaceNames = k8sNodesOtherNamespace.map(node => node.id);
      const k8sNodesOtherNamespaceLinks = links.filter(link => k8sNodesOtherNamespaceNames.includes(link.source) || k8sNodesOtherNamespaceNames.includes(link.target));

      nodes = nodes.filter(nodeGroup => !k8sNodesOtherNamespace.includes(nodeGroup));
      links = links.filter(linkGroup => !k8sNodesOtherNamespaceLinks.includes(linkGroup));
    }

    //Filter by type
    if(this.state.kinds) {
      nodes = nodes.filter(nodeGroup => this.state.kinds.includes(nodeGroup.kind));
    }

    nodesName = nodes.map(nodeGroup => nodeGroup.id);
    links = links.filter(link => nodesName.includes(link.source) && nodesName.includes(link.target));

    const old = new Map(this.state.nodeGroup.data().map(d => [d.id, d]));
    nodes = nodes.map(d => Object.assign(old.get(d.id) || {}, d));
    links = links.map(d => Object.assign({}, d));

    const nodeGroup = this.state.nodeGroup
      .data(nodes, d => d.id)
      .join(enter => enter.append('g')
        .attr('class', 'node-group')
        .attr('id', d => `node-group-${d.id}`))
      .call(this.drag(this.state.simulation));

    const nodeCircle = nodeGroup.append('circle')
      .attr('id', d => `graph-node-${d.id}`)
      .attr('class', d => `graph-node ${d.kind}`)
      .attr('r', this.nodeCircleRadius);

    nodeCircle.append('title')
      .text((d) => { return d.id; });

    const nodeIcon = nodeGroup.append('text')
      .attr('class', d => `fas node-icon ${d.kind}`)
      .attr('text-anchor', 'middle')
      .attr('dominant-baseline', 'middle')
      .attr('font-size', `${this.nodeIconFontSize}px`)
      .text((d) => IconMap[d.kind]);

    const nodeLabel = nodeGroup.append('text')
      .attr('class', 'node-label')
      .classed('hidden', !this.state.showLabels)
      .attr('text-anchor', 'middle')
      .attr('dominant-baseline', 'middle')
      .attr('font-size', `${this.nodeLabelFontSize}px`)
      .attr('y', this.nodeCircleRadius * 2)
      .text((d) => truncate(d.properties.name, this.nodeLabelLength));

    const link = this.state.link
      .data(links, d => [d.source, d.target])
      .join('line')
      .attr('class', 'link');

    nodeGroup
      .on('mouseover', d => this.nodeMouseOver(d3.select(`#node-group-${d.id}`), d))
      .on('mouseout', d => this.nodeMouseOut(d3.select(`#node-group-${d.id}`), d))
      .on('click', d => this.nodeClick(d3.select(`#node-group-${d.id}`), d));

    this.setState({
      nodeCircle: nodeCircle,
      nodeIcon: nodeIcon,
      nodeGroup: nodeGroup,
      showLabels: !nodeLabel.classed('hidden'),
      link: link
    }, () => {
      this.state.simulation.nodes(nodes);
      this.state.simulation.force('link').links(links);
      this.state.simulation.alpha(1).restart();
      this.calculateGraphStats(nodes);
    });
  }

  ticked() {
    this.state.link
      .attr('x1', function (d) { return d.source.x; })
      .attr('y1', function (d) { return d.source.y; })
      .attr('x2', function (d) { return d.target.x; })
      .attr('y2', function (d) { return d.target.y; });

    this.state.nodeGroup
      .attr('transform', d => `translate(${d.x}, ${d.y})`);
  }

  drag(simulation) {
    function dragstarted(d) {
      if (!d3.event.active) simulation.alphaTarget(0.3).restart();
      d.fx = d.x;
      d.fy = d.y;
    }

    function dragged(d) {
      d.fx = d3.event.x;
      d.fy = d3.event.y;
    }

    function dragended(d) {
      if (!d3.event.active) simulation.alphaTarget(0);
      d.fx = null;
      d.fy = null;
    }

    return d3.drag()
      .on('start', dragstarted)
      .on('drag', dragged)
      .on('end', dragended);
  }

  calculateGraphStats(nodes) {
    const stat = {};
    stat.all = nodes.length;
    nodes.reduce((acc, nodeGroup) => {
      stat[nodeGroup.kind] ? stat[nodeGroup.kind] += 1 : stat[nodeGroup.kind] = 1;
      return stat;
    }, stat);

    const statList = [];
    Object.keys(stat).forEach(key => {
      statList.push({
        type: key,
        count: stat[key]
      });
    });

    this.setState({
      stat: statList
    });
  }

  handleStatButton() {
    this.clearClicked();
    this.nodeDetailCard.current.updateNodeData({
      kind: 'Statistics',
      properties: this.state.stat
    }, true);
    this.nodeDetailCard.current.show();
  }

  render() {
    return (
      <div>
        <span className="loader">
          <Loader type="TailSpin" visible={this.state.loading} color='#343a40' />
        </span>
        <Button className="stats" onClick={this.handleStatButton} variant="outline-dark" size="xs">
          Stats
        </Button>
        <div id="chart-area" />
        <NodeDetailCard ref={this.nodeDetailCard} />
        <DateTimePicker onSelect={this.onDateTimeSelect} namespaceOptions={this.state.namespaceOptions}
          objectKindOptions={this.state.objectKindOptions} handleNamespaceChange={this.handleNamespaceChange}
          handleKindChange={this.handleKindChange} showLabels={this.state.showLabels} toggleNodeLabels={this.toggleNodeLabels} 
          defaultKinds={this.state.kinds ? this.state.kinds.map(kind => {
            return {
              label: kind,
              value: kind
            }
          }) : null}/>
      </div>
    );
  }
}
